core/base/validation.hpp

```diff
@@ -0,0 +1,167 @@
+// SPDX-FileCopyrightText: 2017 - 2025 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_CORE_BASE_VALIDATION_HPP_
+#define GKO_CORE_BASE_VALIDATION_HPP_
+
+
+#include <cmath>
+#include <unordered_set>
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/temporary_clone.hpp>
+
+
+namespace gko {
+namespace validation {
+
+
+#define GKO_VALIDATE(_expression, _message)                                 \
+    if (!(_expression)) {                                                   \
+        throw gko::InvalidData(__FILE__, __LINE__, typeid(decltype(*this)), \
+                               _message " (" #_expression ")");             \
+    }
+
+
+template <typename ValueType>
+bool is_finite_scalar(const ValueType& value)
+{
+    return std::isfinite(static_cast<double>(value));
+}
+
+template <typename ValueType>
+bool is_finite_scalar(const std::complex<ValueType>& value)
+{
+    return is_finite_scalar(value.real()) && is_finite_scalar(value.imag());
+}
+
+
+template <typename IndexType>
+bool is_sorted(const gko::array<IndexType>& row_ptrs)
+{
+    const auto host_row_ptrs = row_ptrs.copy_to_host();
+    return std::is_sorted(host_row_ptrs.begin(), host_row_ptrs.end());
+}
+
+
+template <typename IndexType>
+bool is_within_bounds(const gko::array<IndexType>& col_idxs,
+                      const IndexType upper_bound)
+{
+    const auto host_col_idxs = col_idxs.copy_to_host();
+    const auto [min, max] =
+        std::minmax_element(host_col_idxs.begin(), host_col_idxs.end());
+
+    return *min >= 0 && *max < upper_bound;
+}
+
+
+template <typename IndexType>
+bool ell_has_unique_idxs(const gko::array<IndexType>& col_idxs,
+                         const IndexType num_rows, const size_type nnz,
```

Comment:
I would suggest a clearer name for the second parameter
```suggestion
                         const IndexType num_rows, const size_type nnz_per_row,
```